src/util/Scheduler.h

```diff
@@ -0,0 +1,204 @@
+//     async IOs or writing to in-memory buffers) may run them _all_ before
+//     anyone else gets to go, but that's ok precisely because they're very
+//     small actions. The scheduler will shift to other queues exactly when a
+//     queue uses up a _noticable amount of time_ relative to others.
+//
+// This is an old algorithm that was not used for a long time out of fear that
+// it would starve long-duration actions; but it's received renewed study in
+// recent years based on the observation that such starvation only occurs in
+// certain theoretically-tidy but practically-rare distributions of action
+// durations, and the distributions that occur in reality behave quite well
+// under it.
+//
+// The customizations we make are minor:
+//
+//   - We put a floor on the cumulative durations; low cumulative durations
+//     represent a form of "credit" that a queue might use in a burst if it were
+//     to be suddenly full of ready actions, or continuously-reschedule itself,
+//     so we make sure no queue can have less than some (steadily rising) floor.
+//
+//   - We encode deadlines in actions: those with a positive deadline are always
+//     dropped -- unconditionally -- if they are ready to run after their
+//     deadline, and may also be dropped conditionally if the system is under
+//     load / the queue is too long. The sentinel "minimum deadline" value is
+//     reserved to indicate a never-droppable action. To encode a "best effort"
+//     action with no particular deadline that will be dropped only under load,
+//     we set the deadline to the the maximal duration value.
+
+namespace stellar
+{
+
+using Action = std::function<void()>;
+
+class Scheduler
+{
+  public:
+    struct Stats
+    {
+        size_t mActionsEnqueued{0};
+        size_t mActionsDequeued{0};
+        size_t mActionsDroppedDueToOverload{0};
+        size_t mActionsDroppedDueToDeadline{0};
+        size_t mQueuesActivatedFromFresh{0};
+        size_t mQueuesActivatedFromCache{0};
+        size_t mQueuesSuspended{0};
+    };
+
+    using RelativeDeadline = std::chrono::nanoseconds;
+    using AbsoluteDeadline = std::chrono::steady_clock::time_point;
+    static constexpr RelativeDeadline NEVER_DROP = RelativeDeadline::min();
+    static constexpr RelativeDeadline DROP_ONLY_UNDER_LOAD =
+        RelativeDeadline::max();
+    static constexpr AbsoluteDeadline ABS_NEVER_DROP = AbsoluteDeadline::min();
+    static constexpr AbsoluteDeadline ABS_DROP_ONLY_UNDER_LOAD =
+        AbsoluteDeadline::max();
+
+  private:
+    class Queue;
+    using Qptr = std::shared_ptr<Queue>;
+    std::map<std::string, Qptr> mQueues;
+    std::priority_queue<Qptr, std::vector<Qptr>,
+                        std::function<bool(Qptr, Qptr)>>
+        mQueueQueue;
+
+    Stats mStats;
+
+    // A queue is considered "overloaded" if its size is above the load limit.
+    // This is a per-queue limit.
```

Comment:
Expanding on this from a different angle.

At a high level: you can think of all "actions" accumulated in the scheduler as the currently known backlog of work to do (with various deadlines and priorities).

Right now the only thing you can do with this updated design when overloaded is load shed, but this won't help if we continue to add more non droppable work to the backlog. Queues of work will grow without bound or at best will cause processing actions to have some very large latency causing all sorts of problems.

So we need some mechanism to allow the system to backoff and slow down injecting new work into the scheduler if the scheduler is overloaded.

To do that we need to compute an "overloaded" signal that can be used from the *outside* of the scheduler:
* maybe it's as simple as "any queue is overloaded"
* when the scheduler is overloaded, we can then decide on the strategy vs IO
   * in master, I went with "drain the execution queue entirely" but this has a bunch of problems, especially if there are some recurrent tasks (apply bucket type of thing)
   * with this version of the scheduler, I suspect you can instead use a strategy that is just changing the ratio IO vs scheduler by a lot (right now you're giving 1:1 in a later commit).
   * not picking things up from the IO queues increases the chance of dropping connections etc, but that's the trade off we're making here when overloaded. I don't know if there is a good alternative (we could try first to drop some connections, but that seems more complicated).

In terms of the quality of the "overloaded" signal (what my previous comment was discussing), it also probably needs to be a little stickier (this applies to all action types) to avoid bouncing back and forth between "overloaded" and "not overloaded".
The common strategy used in other systems is to use two thresholds: one that causes the queue to transition to "overloaded" and another (much lower) that causes it to transition back to not being overloaded.

Reply:
Implemented a basic version of this using the same queue window-duration (time) parameter for overload detection, and throttling IO (switching from 1:1 to 1:2 ratio) for shedding/backpressure. Doesn't have a two-threshold scheme like you're suggesting, though could possibly be modified in that direction to oscilate less? Unsure.